### DISCUSSION TITLE
Error reading from file/streamError caught: send_ismrmrd_acq failed

### DISCUSSION THREAD

**Discussion by siladokumaci** [gadgetron/gadgetron#1135] [Q&A]
Dear All,

I am very new to Gadgetron and have installed it on my Windows 10 using WSL and Ubuntu-20.04 but am still having issues using it with MATLAB.
My MATLAB (R2018a) is not in WSL but I put the script in WSL and can launch MATLAB from WSL by typing matlab. In addition, gadgetron --info command results in --Matlab Support: YES and I have also downloaded the gadgetron toolbox in MATLAB.

On the MATLAB side, I am running the gadgetron.external.listen(18000, @handle_connection) command and it seems to be listening: Starting ISMRMRD server with handler 'handle_connection' in state: [PASSIVE]
Waiting for connection from client on port: 18000

There is a handle_connection function just as in the tutorial. It is very basic with a disp(''handle connection was called.") in there along with connection.next().

Then when I call the gadgetron_ismrmrd_client from the WSL side as attached, I am getting this send_ismrmrd_acq failed error. The xml file (matlab_tutorial.xml) has the necessary parts for MATLAB in the stream-external as well: connect port="18000" 

Another thing I noticed was, if I use the -C option as in the tutorials, it is showing the default.xml file below as the configuration file and only the -c option is showing the .xml file that I would like to use. I noticed that -C filename.xml in the tutorial was also showing default.xml below for the configuration file. How do we make sure it is using the right file?

I am not sure if my problem is a MATLAB connection issue or something else and would be very happy to hear your suggestions about this.
Sila

![GadgetronError_27072022_asd](https://user-images.githubusercontent.com/45688989/181230099-6ec7c5c9-e6ce-42b9-bcaa-81142c40fdfc.jpg)

**Comment by hansenms** [comment]
Few things to untangle here. Firstly, using `-C` is the correct way to do it. The listing of `default.xml` when you do that is a bit of a bug, it is really because that is the default configuration name, but it is not used when you use `-C`, it will read the local file. Secondly, you also need a Gadgetron running inside the WSL, the connection will not be directly from the client to Matlab, it needs to go to a Gadgetron instance. So do you have that? What do you see in terms of logs there? Are you able to connect to it at all?

**Reply by siladokumaci** [reply]
Thank you for clarifying the -C vs -c. Yes, I am running Gadgetron inside the WSL and I could connect to it while using Python. You may find the error attached.

![GadgetronError2_27072022_asd](https://user-images.githubusercontent.com/45688989/181371016-29435e61-ce4d-4b0b-ab7b-a4367899a5e6.jpg)

**Comment by hansenms** [comment]
So from this, it looks like there is no connection between your WSL environment and Matlab....running in Windows......probably expected. I have not tried this configuration before and I am not totally sure how I would make that work.

**Reply by siladokumaci** [reply]
Thank you very much for your help!
I had followed the "Integrating Matlab running in Windows with Gadgetron in WSL" instructions [here](https://github.com/gadgetron/GadgetronOnlineClass/tree/master/Installation) and the installation instructions [here.](https://github.com/gadgetron/gadgetron/wiki/Using-Matlab-with-Gadgetron#Verification) I hope there is someone who made this configuration work.

**Reply by siladokumaci** [reply]
I just noticed this: Matlab 2018b or newer and I have Matlab 2018a.
Maybe that's the reason? But it will take a bit longer to test this.

**Reply by siladokumaci** [reply]
I downloaded R2022a, the Gadgetron toolbox and modified the path to launch MATLAB R2022a with the matlab command inside WSL. However, I am still getting the same errors, unfortunately.

**Reply by Andrew-Dupuis** [reply]
Both the Gadgetron online class and the Installation instructions you linked to are quite possibly out of date, as they predate the Conda environment installation approach. The current installation instructions for the Gadgetron are available at https://gadgetron.readthedocs.io/en/latest/. 

What version of Gadgetron are you using currently, and how is it installed within WSL?

**Reply by Andrew-Dupuis** [reply]
Actually - looking at your links: have you installed Gadgetron via a package manager, such as apt-get install Gadgetron?

**Reply by siladokumaci** [reply]
Thank you for your help. As far as I remember I first tried a package manager but there were issues whose details I cannot remember anymore, then we tried downloading it from this link [here](https://github.com/gadgetron/gadgetron/releases/tag/v4.1). After your suggestion, I wanted to try the package manager again. I think I deleted some files by mistake and started from scratch installing Ubuntu-20.04 again in WSL. Currently, I cannot make any of the options work anymore - no gadgetron. 

The package manager option is giving me this error: gadgetron: error while loading shared libraries: libismrmrd.so.1.4: cannot open shared object file: No such file or directory
"sudo ldconfig" which I saw in one of the discussions does not help to solve it.

For the option downloading from the link, the last time I had run these commands below and then cmake .. and the rest had worked:
sudo apt-get install build-essential \
git-core wget make cmake gcc-multilib libgtest-dev libboost-all-dev \
libarmadillo-dev libopenblas-dev libfftw3-dev liblapack-dev liblapacke-dev \
libxml2-dev libxslt-dev libpugixml-dev libhdf5-dev libplplot-dev libdcmtk-dev \ python3-dev python3-pip python3-h5py python3-scipy ninja-build \
librocksdb-dev
googletest googletest-tools librange-v3-dev nlohmann-json3-dev libboost-all-dev
However, currently I am getting a "Unable to locate package python3-dev" error which I have not been able to solve yet. 

If I use a Linux machine instead of trying to make it work within WSL, would the Conda environment installation approach that you suggested above be the best? I am not familiar with the Conda environment and had just started using Python while following these tutorials. Will I be able to use the MATLAB as well?

**Reply by hansenms** [reply]
Definitely use the conda installation approach. That is how we manage dependencies and older methods are not really maintained.

**Reply by siladokumaci** [reply]
Thank you both for your help. I will try the Conda installation next week hopefully. Also if it is not very difficult to do, perhaps it could be helpful to make new users aware of these installation difficulties within WSL by updating those pages and directing to the latest suggested installation.

**Reply by hansenms** [reply]
We should archive that online course repo. It is not being maintained.

**Reply by siladokumaci** [reply]
Hi again,
 
I tried the Gadgetron installation in the conda environment on Ubuntu [from this link](https://gadgetron.readthedocs.io/en/latest/building.html) that you suggested before and all end-to-end tests passed following the command: python run_tests.py --echo-log-on-failure --timeout=600 -F --stats stats.csv cases/*

Then I tried the MATLAB test (./run_tests.py cases/external_matlab_acquisition_example.cfg). 
Currently there is MATLAB support and I could make it work in a previous installation but now I am getting this ModuleNotFoundError for run_gadgetron_test.py line 25 that there is no module named ismrmrd and I am not sure where it is looking for this to import. I installed the ismrmrd by following the instructions [here](https://github.com/ismrmrd/siemens_to_ismrmrd). I also tried copying the ismrmrd folder to /usr/local/ and set the ISMRMRD environmental variable as in the instructions which did not help with this error above. 

The current arrangement of the folders is that I have an anaconda3 folder under which there are gadgetron, ismrmrd, and envs folders (envs also has gadgetron). The installation for siemens_to_ismrmrd also gave some errors but in general I am not sure where to install ISMRMD and siemens_to_ismrmrd in anaconda3 as I am not familiar with any of these and would be very grateful for your help.
